### ISSUE TITLE
[Mixin] Handle @Dynamic

### ISSUE THREAD

**Issue by DenWav** [DemonWav/MinecraftDevIntelliJ#298]
Forge remaps all `System.exit()` calls to `Lnet/minecraftforge/fml/common/asm/transformers/TerminalTransformer$ExitVisitor;systemExitCalled(I)V`, which should be handled by MCDev whenever Mixing into Forge sources. Currently it's erroring in these cases.

**Comment by Mumfrey** [comment]
One thing I thought was cool when you were showing me IntelliJ stuff last night was that you could configure exceptions on a per-inspection basis. Would you be able to have this inspection have similar functionality so that when the plugin says *"this method doesn't exist"* one of the quick-fixes is like an *"I know this will exist at run-time"*? You could preconfigure certain exceptions like the `System.exit` one mentioned, but could also provide functionality for the user to define their own exceptions to the rule too maybe?

**Comment by DenWav** [comment]
That's a good idea, I like it.

**Comment by stephan-gh** [comment]
I personally think it would be better if we have a way to mark them as runtime-only directly in the source code. This could be either with an option directly in Mixin (might be possible to use it for validation in the annotation processor as well), or some kind of special comment that is only recognized by MinecraftDev.

Using inspection exceptions for this is cool, but everyone needs to apply them manually, even though it's actually something you know when creating the injection. You could still have a quick fix that marks them properly in the source code.

**Comment by Mumfrey** [comment]
That's a good point, I don't mind adding some decoration in mixin itself which could be used for this. As you say, it helps when anyone looks at the code then and actually lets them know that the target won't be visible when looking at the original source of the target. Case in point: the [MixinWorldServer mixin which always confuses people because it mixes in to an upstream mixin](https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/mixin/tileentityactivation/MixinWorldServer_TileEntityActivation.java#L43).

**Comment by Mumfrey** [comment]
So after our discussion last night, I will go ahead and add an `@Dynamic` or similar annotation which can be used to decorate handler methods whose targets are not available in the original class (eg. are added by a transformer or upstream mixin) which will be something of the form:

```java
@Dynamic("System.exit is transformed by FML TerminalTransformer.")
@Inject(method = "shutdownMinecraftApplet", at = @At(value = "INVOKE", target = FORGE_TRANSFORMER_EXIT_VISITOR, remap = false))
public void onShutdownDelegate(CallbackInfo ci) {
    SpongeImpl.postShutdownEvents();
}
```

I'd really like to use `@Runtime` but I feel like it's irresponsible to use the name of a core Java class. Thoughts?

**Comment by stephan-gh** [comment]
`@Dynamic` sounds fine to me. Should that disable warnings/errors for all targets that are mentioned in there? In your example, `shutdownMinecraftApplet` exists at compile time and only Forge's exit method would need `@Dynamic`. Ideally, we could flag a specific target (e.g. only the injection point or only the target method) as dynamic, but I can't think of any nice way to add this with an annotation.

Now that I think about it, it doesn't really matter in reality, at least if you set the `require` option:  you'd be still notified of the problem at runtime when the mixin is applied.

**Comment by Mumfrey** [comment]
@Minecrell exactly, and that was the gist of the conversation last night. The only way to handle it nicely at the annotation level would be to add a decorator argument to *every* annotation, which would frankly suck from a readability standpoint and completely undermine the purpose of making it more obvious.

My thoughts were that you really only care about this at the *method* level, and all you're doing is saying that

1. Any failure of the IDE to lookup either the target method or injection point can be demoted or ignored
2. Anyone reading the code can knows what's going on
3. At runtime, failure to resolve target exception message can be decorated with the info in the `@Dynamic` annotation

So basically having it at the injection-point level wouldn't achieve anything particularly useful anyway, it might as well just be a method-level decoration.

**Comment by Mumfrey** [comment]
So one possible enhancement to this is that if a mixin member is targetting something known to be added by an upstream mixin, it's certainly feasible to simply use the fully-qualified name of the upstream mixin class as the annotation content. I have noted this [in the Javadoc for the `@Dynamic` annotation](https://github.com/SpongePowered/Mixin/blob/feature/dynamic/src/main/java/org/spongepowered/asm/mixin/Dynamic.java#L73-L76).

I'm imagining that this could provide some scope for allowing this situation:

> Mixin A targets a method added by Mixin B in class C. Method in Mixin A appears to target something which doesn't exist. MCDev has to resolve all mixins in scope which target C, it resolves B and finds the method, this allows for auto-completion of (for example) `@At` parameters on the injector in Mixin A.

From what Kyle was saying, this intermediate step of resolving mixins which target C could potentially be expensive. However if the user pre-decorates the method with `@Dynamic("name.of.UpstreamMixin")` then the resolver can quickly check if the value in `@Dynamic` is a valid class name and assume it to be the upstream contributing mixin, cutting out the step above where MCDev has to resolve all mixins in scope which target class C.

This also presents some opportunities for auto-completion in the `@Dynamic` annotation itself.

**Comment by stephan-gh** [comment]
@Mumfrey You suggested adding `@Dynamic(mixin = some.upstream.Mixin.class)` on IRC recently; I'm not sure if you've seen my reply on IRC so I'll just post it here as well:

If we have `@Dynamic(mixin = /* ... */)` I don't see a reason that we should also support using full qualified class names as string value (at least in MCDev): If you're not able to reference the class directly then we're unable to provide additional assistance anyway (because we can't inspect the Mixin class either).

I'd personally like to avoid having to check if a string is a full qualified name of a class and implementing auto-completion for that manually so I'd prefer having the extra `mixin` parameter.

**Comment by Mumfrey** [comment]
@Minecrell I added it [last week](https://github.com/SpongePowered/Mixin/commit/ce2ea22416f183312c3f27fc876dfb7175bb6c94) I just forgot to push it.

As I noted [in the Javadoc](http://jenkins.liteloader.com/view/Other/job/Mixin/javadoc/index.html?org/spongepowered/asm/mixin/Dynamic.html) the purpose of fully-qualifying the upstream mixin as a string is not for **your** benefit, but the benefit of the person reading the code. Because even if the code can't reference the mixin dynamically, it doesn't mean the mixin isn't in the user's workspace or accessible on github.

In eclipse at least (so I would assume it's the same in IntelliJ), any string which is actually a class name can be ctrl-clicked just like a normal member can, as long as the class exists _somewhere_ in the workspace (if it's fully-qualified it just means that if there is more than one class with that name that it's not necessary to choose from a dropdown). For example, let's say that a user is writing two mods which don't reference each other but the second contains injections for mixins in the first, even though the user _can't_ use `mixin=` they can still put the mixin name as a string in order that it be clickable *in their workspace*, not necessarily for everyone.

For this reason I have chosen to **retain** the note in the javadoc about using the fq name in those situations because I think it's a perfectly valid use case, and not everything has to directly benefit users of mcdev.

> I'd personally like to avoid having to check if a string is a full qualified name of a class and implementing auto-completion for that manually so I'd prefer having the extra mixin parameter.

That's up to you, but I'm still keeping the note, because as I described above, it's still navigable even if mcdev ignores it.

**Comment by Mumfrey** [comment]
Also 0.7.4 has been merged to master now, so it's available for use.

**Comment by DenWav** [comment]
> not everything has to directly benefit users of mcdev

Blasphemy.

**Comment by Mumfrey** [comment]
Sweet
